src/main/java/com/techsophy/tsf/runtime/form/service/impl/FormValidationServiceImpl.java

```diff
@@ -0,0 +1,162 @@
+                    }
+                }
+                else if(component.getType().equals("datagrid")||component.getType().equals("editgrid"))
+                {
+                    if(data.get(component.getLabel())!=null)
+                    {
+                        List<Map<String,Object>> dataGridList= (List<Map<String, Object>>) data.get(component.getLabel());
+                        List<Component> componentList=component.getComponents();
+                        for(Map m:dataGridList)
+                        {
+                            internalLoop(prefix,component,validationResultList,componentList,m,formId);
+                        }
+                    }
+                }
+                else
+                {
+//                    for(Component c:component.getComponents())
+//                    {
+//                        validationResultList.addAll(validateComponent(prefix,c,component.getData(data,new HashMap<>()),formId));
+//                    }
+                    internalLoop(prefix,component,validationResultList,component.getComponents(),data,formId);
+                }
+            }
+            if(component.getType().equals("datamap"))
+            {
+                validationResultList.addAll(validateComponent(prefix,component.getValueComponent(),component.getData(data,component.getValueComponent().getType(),component.getLabel()),formId));
+            }
+            List<Columns> columns=component.getColumns();
+            if(columns!=null)
+            {
+                List<Component> componentList=new ArrayList<>();
+                for(Columns c:columns)
+                {
+                    componentList.addAll(c.getComponent());
+                }
+//               List<Component> componentList= columns.stream().map(x->x.getComponent().get(0)).collect(Collectors.toList());
+//               for (Component value : c)
+//                {
+//                    validationResultList.addAll(validateComponent(prefix, value, data, formId));
+//                }
+                internalLoop(prefix,component,validationResultList,componentList,data,formId);
+            }
+            if(component.getRows()!=null)
+            {
+                List<Component> componentList= component.getRows().stream().flatMap(x -> x.stream().map(Internal::getComponents)).flatMap(Collection::stream).collect(Collectors.toList());
+//                for(Component component1:componentList)
+//                {
+//                   validationResultList.addAll(validateComponent(prefix,component1,data,formId));
+//                }
+                internalLoop(prefix,component,validationResultList,componentList,data,formId);
+            }
+        }
+        else
+        {
+            if(component.getValidate()!=null)
+            {
+                return component.getValidate().validate(new Validate.ComponentData(component,data,prefix,formId),globalMessageSource,mongoTemplate);
```

Comment:
if you are not able to follow the approach discuss with me, this whole class needs to be refactored